R/packages.R

```diff
@@ -0,0 +1,282 @@
+      metadata$assets <- get_github_wasm_assets(desc)
+      metadata$type <- "library"
+    } else if (grepl("r-universe\\.dev$", repo)) {
+      metadata$assets <- get_r_universe_wasm_assets(desc)
+      metadata$type <- "package"
+    } else {
+      # Fallback to repo.r-wasm.org lookup for CRAN and anything else
+      metadata$assets <- get_default_wasm_assets(desc)
+      metadata$type <- "package"
+    }
+  } else if (verbose) {
+    message("Skipping cached Wasm binary: ", metadata$ref)
+  }
+
+  metadata
+}
+
+download_wasm_packages <- function(appdir, destdir, verbose, package_cache) {
+  verbose_print <- if (verbose) message else list
+  # App dependencies, ignoring shiny packages in base webR image
+  pkgs <- unique(renv::dependencies(appdir, quiet = !verbose)$Package)
+  pkgs <- pkgs[pkgs != "shiny" & pkgs != "bslib"]
+  if (length(pkgs) > 0) {
+    pkgs <- resolve_dependencies(pkgs, verbose)
+  }
+
+  if (verbose) {
+    p <- progress::progress_bar$new(
+      format = "[:bar] :percent\n",
+      total = length(pkgs),
+      clear = TRUE
+    )
+  }
+
+  # Create empty R packages directory in app assets if not already there
+  pkg_dir <- fs::path(destdir, "shinylive", "webr", "packages")
+  if (!fs::dir_exists(pkg_dir)) {
+    verbose_print("Creating ", pkg_dir, "/")
+    fs::dir_create(pkg_dir)
+  }
+
+  verbose_print(
+    "Downloading WebAssembly R package binaries to ", pkg_dir, "/"
+  )
+
+  # Load existing metadata from disk, from a previously deployed app
+  metadata_file <- fs::path(destdir, "shinylive", "webr", "packages", "metadata.rds")
+  prev_metadata <- if (package_cache && fs::file_exists(metadata_file)) {
+    readRDS(metadata_file)
+  } else {
+    list()
+  }
+
+  # Loop over packages and download them if not cached
+  names(pkgs) <- pkgs
+  cur_metadata <- sapply(pkgs, function(pkg) {
+    if (verbose) p$tick()
+
+    pkg_subdir <- fs::path(pkg_dir, pkg)
+    if (!fs::dir_exists(pkg_subdir)) {
+      fs::dir_create(pkg_subdir)
+    }
+
+    prev_meta <- if (pkg %in% names(prev_metadata)) {
+      prev_metadata[[pkg]]
+    } else {
+      list()
+    }
+    # Create package ref and lookup download URLs
+    meta <- prepare_wasm_metadata(pkg, prev_meta, verbose)
+
+    if (!meta$cached && length(meta$assets) > 0) {
+      # Download Wasm binaries and copy to static assets dir
+      for (file in meta$assets) {
+        utils::download.file(file$url, fs::path(pkg_subdir, file$filename))
+      }
+      meta$cached <- TRUE
+      meta$path <- glue::glue("packages/{pkg}/{meta$assets[[1]]$filename}")
+    }
+    meta
+  }, simplify = FALSE, USE.NAMES = TRUE)
+
+  # Merge metadata to protect previous cache
+  pkgs <- unique(c(names(prev_metadata), names(cur_metadata)))
+  metadata <- mapply(
+    function(a, b) if (is.null(b)) a else b,
+    prev_metadata[pkgs],
+    cur_metadata[pkgs],
+    SIMPLIFY = FALSE, USE.NAMES = FALSE
+  )
+  names(metadata) <- pkgs
```

Comment:
Since we're setting the names directly, we can ignore the parameter. We can also use `Map` instead of `mapply(SIMPLIFY=FALSE)`
```suggestion
  metadata <- Map(
    function(a, b) if (is.null(b)) a else b,
    prev_metadata[pkgs],
    cur_metadata[pkgs]
  )
  names(metadata) <- pkgs
```